File: ipd.scala
object Main {
  sealed trait Decision extends Product with Serializable
  case object Cooperate extends Decision
  case object Defect extends Decision

  type Trial = (Decision, Decision)
  type History = List[Decision]
  type Prisoners = List[Prisoner]

  case class Prisoner(name: String,
                      strategy: Strategy,
                      history: History = List.empty,
                      totalScore: Int = 0)

  trait Strategy {
    def move(mover: History, opponent: History): History
    def cooperate(history: History) = Cooperate :: history
    def defect(history: History) = Defect :: history
  }

  case object AlwaysCooperate extends Strategy {
    def move(mover: History, opponent: History): History = cooperate(mover)
  }

  case object AlwaysDefect extends Strategy {
    def move(mover: History, opponent: History): History = defect(mover)
  }

  def score(trial: Trial): Int = trial match {
    case (Cooperate, Defect) => 0
    case (Defect, Defect) => 1
    case (Cooperate, Cooperate) => 3
    case (Defect, Cooperate) => 5
  }

  def game(prisoners: Prisoners) = {
    prisoners.zip(prisoners.reverse).map { case (mover, opponent) =>
      val strategy = mover.strategy
      mover.copy(history = strategy.move(mover.history, opponent.history))
    }
  }

  def iteratedGame(iterations: Int, prisoners: Prisoners): Prisoners = {
    1.to(iterations).foldLeft(prisoners){ (prisoners, _) => game(prisoners) }
  }

  def totalScores(prisoners: Prisoners): Prisoners = {
    prisoners.zip(prisoners.reverse).map { case (mover, opponent) =>
      val totalScore = mover.history.zip(opponent.history).foldLeft(0) { case (total, (m, o)) =>
        total + score((m, o))
      }
      mover.copy(totalScore = totalScore)
    }
  }

  def main(args: Array[String]): Unit = {
    val prisoners = List(Prisoner("a", AlwaysCooperate), Prisoner("b", AlwaysDefect))

    totalScores(iteratedGame(10, prisoners)).foreach { p =>
      println(s"Prisoner ${p.name.toUpperCase} (${p.strategy}): ${p.totalScore}")
    }
  }
}

File: ipd.rb
require 'hamster/list'
require 'hamster/hash'

module Prisoner
  def prisoner(name, strategy)
    Hamster.hash(name: name, strategy: strategy, moves: Hamster.list, total: 0)
  end
end

module Strategy
  def strategy(strategy)
    -> mover, other { strategy.(mover[:moves], other[:moves]) }
  end

  def always_cooperate(mover, other); cooperate(mover); end

  def always_defect(mover, other); defect(mover); end
end

module Moves
  def cooperate(moves); 1; end
  def defect(moves); 0; end
end

module Game
  def to_both(prisoners)
    prisoners.zip(prisoners.reverse).map do |mover, opp|
      yield mover, opp
    end
  end

  def game(prisoners)
    to_both(prisoners) do |mover, opp|
      move = mover[:strategy].(mover, opp)
      mover.put(:moves, mover[:moves].cons(move))
    end
  end

  def iterated_game(iterations, prisoners)
    (1..iterations).reduce(prisoners) do |prisoners, _|
      total_score(game(prisoners))
    end
  end

  def total_score(prisoners)
    to_both(prisoners) do |mover, opp|
      score = mover[:moves].zip(opp[:moves]).reduce(0) do |score, (m, o)|
        score += compute_score(m, o)
      end
      mover.put(:total, score)
    end
  end

  def compute_score(a, b)
    @scores ||= {
      [0, 1] => 5,
      [1, 1] => 3,
      [0, 0] => 1,
      [1, 0] => 0
    }

    @scores[[a,b]]
  end
end

[Prisoner, Moves, Strategy, Game].each { |m| include m }

prisoners = iterated_game(10, Hamster.list(
  prisoner(:a, strategy(method(:always_cooperate))),
  prisoner(:b, strategy(method(:always_defect)))
))

puts "Prisoner A (AlwaysCooperate): %d" % prisoners.head[:total]
puts "Prisoner B (AlwaysDefect)   : %d" % prisoners.last[:total]
